### DISCUSSION TITLE
Seeking Project Idea to Implement ASP.NET Core Web APIs

### DISCUSSION THREAD

**Discussion by phadkesharan** [dotnet/aspnetcore#55725] [General]
Hello fellow developers,

I've recently completed all the Microsoft Learn courses related to ASP.NET Core and have delved into the basics of Clean Architecture by Ardalis. Additionally, I have a solid proficiency in developing Web APIs. Now, I'm on the lookout for a project idea where I can put my knowledge and skills into practice, primarily focusing on Web APIs while also considering the integration of SPAs or MVC.

Here are a few key points to consider:

ASP.NET Core Web APIs: The project should primarily revolve around the development of Web APIs using ASP.NET Core, ensuring adherence to best practices for creating scalable, maintainable, and secure APIs.

Clean Architecture: I aim to implement Clean Architecture principles, separating concerns and maintaining a clear separation of layers, such as domain logic, application logic, and infrastructure, to promote flexibility and testability.

Integration with SPA or MVC: While the main focus is on Web APIs, I'm open to integrating a Single Page Application (SPA) or MVC components to enhance the user experience and showcase full-stack development capabilities.

Functional Requirements: Detailed functional requirements or use cases would greatly assist in shaping the project idea and ensuring it aligns with real-world scenarios.

I'm eager to hear your suggestions, insights, and any references you can provide. Let's collaborate to create a meaningful project that not only hones my skills but also contributes value to the developer community.

Thank you in advance for your contributions!

**Comment by davidfowl** [comment]
https://www.reddit.com/r/dotnet/ might be a better place to farm ideas.

**Comment by markmouton22** [comment]
Hi @phadkesharan :)

How about the old classic, build a point of sales system?

**Reply by markmouton22** [reply]
I asked co pilot to help with some of the admin of an high-level overview:


Here’s a high-level overview of what the system could include:

**1. Product Catalog:**
API endpoints for CRUD operations on products.

Each product can have properties like ID, name, description, price, and quantity.

**2. Sales Transactions:**

API endpoints to create a new sale, add products to the sale, and complete the sale.
Each sale can have multiple products, each with a quantity. The total cost of the sale is the sum of the cost of each product times its quantity.

**3. Inventory Management:**

API endpoints to update product quantities.
When a sale is completed, the quantities of the products in the sale are deducted from the inventory.
Here’s a high-level overview of what the system could include:

**1. Product Catalog:**

API endpoints for CRUD operations on products.
Each product can have properties like ID, name, description, price, and quantity.

**2. Sales Transactions:**

API endpoints to create a new sale, add products to the sale, and complete the sale.
Each sale can have multiple products, each with a quantity. The total cost of the sale is the sum of the cost of each product times its quantity.

**3. Inventory Management:**

API endpoints to update product quantities.
When a sale is completed, the quantities of the products in the sale are deducted from the inventory.

**4. User Management:**

API endpoints for user registration, user login, and user roles.
Different roles (like cashier, manager) can have different permissions.
**5. Reports:**

API endpoints to generate reports, like total sales for a given period, most popular products, etc.

**If you decide to integrate a SPA or MVC:**

You can create views for users to interact with the system, like a product catalog view, a sales view, etc.
Remember to follow Clean Architecture principles when designing your system. For example, the business logic (like calculating the total cost of a sale) should be in the domain layer, separate from the API endpoints in the application layer and the database code in the infrastructure layer.

API endpoints for user registration, user login, and user roles.
Different roles (like cashier, manager) can have different permissions.

@phadkesharan  happy to help when you start
